File: frontend/lib/analysis.ts
const API_BASE_URL =
  process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";

export type RiskCategory =
  | "Market Risk"
  | "Credit Risk"
  | "Operational Risk"
  | "Regulatory/Compliance Risk"
  | "Strategic Risk"
  | "Reputational Risk";

/** A single risk factor gathered from Item 1A, tagged with a selection id. */
export type RiskFactor = {
  id: number;
  title: string;
  summary: string | null;
  category: RiskCategory | null;
  verbatim_text: string;
};

/** One assessed risk factor (stub output until real analysis lands). */
export type Assessment = {
  title: string;
  category: RiskCategory | null;
  status: string;
  note: string | null;
};

export type StartResponse = {
  thread_id: string;
  current_year_risk_factors: RiskFactor[];
};

export type SelectResponse = {
  thread_id: string;
  assessment: Assessment[];
};

async function parseError(res: Response): Promise<string> {
  try {
    const body = (await res.json()) as { detail?: unknown };
    if (typeof body?.detail === "string") return body.detail;
  } catch {
    // fall through to the status-based message
  }
  return `Request failed: ${res.status}`;
}

/**
 * Start a run. The workflow gathers every risk factor, then pauses; the
 * returned thread_id identifies the paused run for selection and resume.
 */
export async function startAnalysis(input: {
  cik: string;
  currentYear: number;
  priorYear: number;
}): Promise<StartResponse> {
  const res = await fetch(new URL("/analysis/start", API_BASE_URL), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      cik: input.cik,
      current_year: input.currentYear,
      prior_year: input.priorYear,
    }),
  });
  if (!res.ok) throw new Error(await parseError(res));
  return (await res.json()) as StartResponse;
}

/** Re-fetch a paused run's gathered risk factors (survives page refresh). */
export async function getAnalysis(threadId: string): Promise<StartResponse> {
  const res = await fetch(
    new URL(`/analysis/${encodeURIComponent(threadId)}`, API_BASE_URL),
  );
  if (!res.ok) throw new Error(await parseError(res));
  return (await res.json()) as StartResponse;
}

/** Resume a paused run with the chosen factor ids and return the assessment. */
export async function submitSelection(
  threadId: string,
  selectedIds: number[],
): Promise<SelectResponse> {
  const res = await fetch(
    new URL(`/analysis/${encodeURIComponent(threadId)}/select`, API_BASE_URL),
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ selected_ids: selectedIds }),
    },
  );
  if (!res.ok) throw new Error(await parseError(res));
  return (await res.json()) as SelectResponse;
}

File: backend/app/api/routes/analysis.py
"""Analysis routes: start a risk-intelligence run, then resume it with the
user's chosen risk factors.

The LangGraph workflow (``agent.agent.graph``) gathers every risk factor from a
company's 10-K, then pauses at a human-in-the-loop interrupt. ``/analysis/start``
drives the graph to that pause and returns the factors; the user picks the ones
worth assessing; ``/analysis/{thread_id}/select`` resumes the same run (via the
in-memory checkpointer, keyed by ``thread_id``) through the assessment stage.
"""

from typing import Annotated, Any
from uuid import uuid4

from fastapi import APIRouter, HTTPException, Path
from langgraph.types import Command
from pydantic import BaseModel, Field

from agent.agent import graph
from agent.state import RiskCategory, WorkflowState

router = APIRouter(prefix="/analysis", tags=["analysis"])


class StartRequest(BaseModel):
    """Kick off a run for one company across two fiscal years."""

    cik: str = Field(description="SEC Central Index Key of the company.")
    current_year: int = Field(description="Fiscal year of the 10-K to analyze.")
    prior_year: int = Field(description="Fiscal year of the 10-K to compare against.")


class RiskFactorOut(BaseModel):
    """A gathered risk factor, tagged with a stable id for selection."""

    id: int = Field(description="Index of this factor within the gathered list.")
    title: str
    summary: str | None = None
    category: RiskCategory | None = None
    verbatim_text: str


class StartResponse(BaseModel):
    """Risk factors gathered before the workflow paused for selection."""

    thread_id: str
    current_year_risk_factors: list[RiskFactorOut]


class SelectRequest(BaseModel):
    """The user's chosen subset, by the ids returned from ``/start``."""

    selected_ids: list[int] = Field(
        description="Ids of the risk factors to assess in depth."
    )


class AssessmentRow(BaseModel):
    """One assessed risk factor (stub output for now)."""

    title: str
    category: RiskCategory | None = None
    status: str
    note: str | None = None


class SelectResponse(BaseModel):
    """Assessment produced after resuming the run with the selected factors."""

    thread_id: str
    assessment: list[AssessmentRow]


def _thread_config(thread_id: str) -> dict[str, Any]:
    return {"configurable": {"thread_id": thread_id}}


def _to_out(factors: list[dict[str, Any]]) -> list[RiskFactorOut]:
    """Attach an index-based id to each gathered factor for stable selection."""
    return [
        RiskFactorOut(
            id=i,
            title=factor.get("title", ""),
            summary=factor.get("summary"),
            category=factor.get("category"),
            verbatim_text=factor.get("verbatim_text", ""),
        )
        for i, factor in enumerate(factors)
    ]


@router.post("/start", response_model=StartResponse)
async def start_analysis(request: StartRequest) -> StartResponse:
    """Run the workflow up to the selection pause and return the risk factors."""
    thread_id = str(uuid4())
    config = _thread_config(thread_id)
    initial_state: WorkflowState = {
        "company_cik": request.cik,
        "current_year": request.current_year,
        "prior_year": request.prior_year,
    }

    result = await graph.ainvoke(initial_state, config)
    factors = result.get("current_year_risk_factors", [])
    return StartResponse(
        thread_id=thread_id,
        current_year_risk_factors=_to_out(factors),
    )


async def _gathered_factors(thread_id: str) -> list[dict[str, Any]]:
    """Return the gathered current-year factors for a thread, or 404."""
    snapshot = await graph.aget_state(_thread_config(thread_id))
    factors = snapshot.values.get("current_year_risk_factors")
    if not snapshot.values or factors is None:
        raise HTTPException(
            status_code=404,
            detail=(
                f"No analysis found for thread {thread_id!r}. It may have "
                "expired (state is held in memory and is lost on restart)."
            ),
        )
    return factors


@router.get("/{thread_id}", response_model=StartResponse)
async def get_analysis(
    thread_id: Annotated[str, Path(description="Thread id from /analysis/start.")],
) -> StartResponse:
    """Re-fetch the gathered risk factors so the selection page survives refresh."""
    factors = await _gathered_factors(thread_id)
    return StartResponse(
        thread_id=thread_id,
        current_year_risk_factors=_to_out(factors),
    )


@router.post("/{thread_id}/select", response_model=SelectResponse)
async def select_risk_factors(
    thread_id: Annotated[str, Path(description="Thread id from /analysis/start.")],
    request: SelectRequest,
) -> SelectResponse:
    """Resume the run with the chosen factors and return the assessment."""
    factors = await _gathered_factors(thread_id)

    valid_ids = range(len(factors))
    invalid = [i for i in request.selected_ids if i not in valid_ids]
    if invalid:
        raise HTTPException(
            status_code=400,
            detail=f"Unknown risk factor ids for this analysis: {invalid}.",
        )
    if not request.selected_ids:
        raise HTTPException(
            status_code=400, detail="Select at least one risk factor to assess."
        )

    selected = [factors[i] for i in request.selected_ids]
    result = await graph.ainvoke(Command(resume=selected), _thread_config(thread_id))
    return SelectResponse(
        thread_id=thread_id,
        assessment=[AssessmentRow(**row) for row in result.get("assessment", [])],
    )
